src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -552,41 +556,141 @@ def calculate_dense(self):
     def expectation(self, state, normalize=False):
         return Hamiltonian.expectation(self, state, normalize)
 
-    def expectation_from_samples(self, freq, qubit_map=None):
-        terms = self.terms
-        for term in terms:
+    def _exp_from_circuit(
+        self, circuit: "Circuit", qubit_map: dict, nshots: int = None
+    ):
+        """
+        Calculate the expectation value from a circuit.
+        This allows for non diagonal observables. Each term of the observable is
+        treated separately, by measuring in the correct basis and re-executing the
+        circuit.
+
+        Args:
+            circuit (Circuit): input frequencies.
+            qubit_map (dict): qubit map.
+
+        Returns:
+            (float): the calculated expectation value.
+        """
+        from qibo import Circuit, gates
```

Comment:
Is this causing a circular import if top level?

Reply:
Unfortunately, it is.

Reply:
Thanks for confirming. We could investigate if it is easy to fix this, for sure not in this PR, but maybe open an issue.